### PR TITLE
Added output type s3 

### PR DESCRIPTION
**Pull request by evalsocket**
Signed-off-by: Yuvraj <evalsocket@gmail.com>

**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

 /area outputs

**What this PR does / why we need it**:
Added new output type s3. Now you can upload json payload to s3 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: outputs/aws.go [hunk 1]
```diff
@@ -135,6 +138,37 @@ func (c *Client) SendMessage(falcopayload types.FalcoPayload) {
 	c.PromStats.Outputs.With(map[string]string{"destination": "awssqs", "status": "ok"}).Inc()
 }
 
+// UploadS3 upload payload to S3
+func (c *Client) UploadS3(falcopayload types.FalcoPayload) {
+	f, _ := json.Marshal(falcopayload)
+
+	prefix := ""
+	if c.Config.AWS.S3.Prefix != "" {
```
Comment (Issif) [inline outputs/aws.go:146]: You need to declare the config https://github.com/falcosecurity/falcosidekick/blob/master/config.go

File: outputs/aws.go [hunk 2]
```diff
@@ -135,6 +138,37 @@ func (c *Client) SendMessage(falcopayload types.FalcoPayload) {
...
+	}
+
+	key := fmt.Sprintf("%s/payload-%s-%v.json", falcopayload.Rule, prefix, time.Now().UnixNano())
+	resp, err := s3.New(c.AWSSession).PutObject(&s3.PutObjectInput{
+		Bucket: aws.String(c.Config.AWS.S3.Bucket),
+		Key:    aws.String(key),
+		Body:   bytes.NewReader(f),
+	})
+	if err != nil {
+		go c.CountMetric("outputs", 1, []string{"output:awss3", "status:error"})
```
Comment (Issif) [inline outputs/aws.go:157]: you need to declare the stats in https://github.com/falcosecurity/falcosidekick/blob/master/stats.go

File: outputs/aws.go [hunk 3]
```diff
@@ -135,6 +138,37 @@ func (c *Client) SendMessage(falcopayload types.FalcoPayload) {
 	c.PromStats.Outputs.With(map[string]string{"destination": "awssqs", "status": "ok"}).Inc()
 }
 
+// UploadS3 upload payload to S3
+func (c *Client) UploadS3(falcopayload types.FalcoPayload) {
```
Comment (Issif) [inline outputs/aws.go:142]: You need to declare the new client in https://github.com/falcosecurity/falcosidekick/blob/master/main.go

You need to declare the handler in https://github.com/falcosecurity/falcosidekick/blob/master/handlers.go

**Review by Issif — CHANGES_REQUESTED** [review]

File: config.go [hunk 4]
```diff
@@ -91,6 +91,9 @@ func getConfig() *types.Configuration {
 	v.SetDefault("AWS.CloudWatchLogs.LogGroup", "")
 	v.SetDefault("AWS.CloudWatchLogs.LogStream", "")
 	v.SetDefault("AWS.CloudWatchLogs.MinimumPriority", "")
+	v.SetDefault("AWS.S3.MinimumPriority", "")
+	v.SetDefault("AWS.S3.Bucket", "")
+	v.SetDefault("AWS.S3.Prefix", "falcosidekick")
```
Comment (Issif) [inline config.go:96]:
```suggestion
	v.SetDefault("AWS.S3.Bucket", "")
	v.SetDefault("AWS.S3.Prefix", "falco")
	v.SetDefault("AWS.S3.MinimumPriority", "")
```

I prefer to keep same order than others and it's more logical to use `falco` as default prefix.

File: handlers.go [hunk 5]
```diff
@@ -174,13 +174,17 @@ func forwardEvent(falcopayload types.FalcoPayload) {
 	}
 
 	if config.AWS.SNS.TopicArn != "" && (falcopayload.Priority >= types.Priority(config.AWS.SNS.MinimumPriority) || falcopayload.Rule == testRule) {
-		go awsClient.PublishTopic(falcopayload)
+		go awsClient.UploadS3(falcopayload)
```
Comment (Issif) [inline handlers.go:177]:
```suggestion
		go awsClient.PublishTopic(falcopayload)
```

File: handlers.go [hunk 6]
```diff
@@ -174,13 +174,17 @@ func forwardEvent(falcopayload types.FalcoPayload) {
...
 	}
 
 	if config.AWS.CloudWatchLogs.LogGroup != "" && (falcopayload.Priority >= types.Priority(config.AWS.CloudWatchLogs.MinimumPriority) || falcopayload.Rule == testRule) {
 		go awsClient.SendCloudWatchLog(falcopayload)
 	}
 
+	if config.AWS.S3.Bucket != "" && (falcopayload.Priority >= types.Priority(config.AWS.S3.MinimumPriority) || falcopayload.Rule == testRule) {
+		go awsClient.PublishTopic(falcopayload)
```
Comment (Issif) [inline handlers.go:185]:
```suggestion
		go awsClient.UploadS3(falcopayload)
```

File: main.go [hunk 7]
```diff
@@ -221,6 +221,7 @@ func init() {
 			config.AWS.Region = ""
 			config.AWS.Lambda.FunctionName = ""
 			config.AWS.SQS.URL = ""
+			config.AWS.S3.Bucket = ""
```
Comment (Issif) [inline main.go:224]: For line #214 (I can't propose on a suggestion for sorry)

```go
if config.AWS.Lambda.FunctionName != "" || config.AWS.SQS.URL != "" ||
		config.AWS.SNS.TopicArn != "" || config.AWS.CloudWatchLogs.LogGroup != "" || config.AWS.S3.Bucket != ""
```

File: outputs/aws.go [hunk 8]
```diff
@@ -135,6 +138,38 @@ func (c *Client) SendMessage(falcopayload types.FalcoPayload) {
...
+	f, _ := json.Marshal(falcopayload)
+
+	prefix := ""
+	t := time.Now()
+	if c.Config.AWS.S3.Prefix != "" {
+		prefix = c.Config.AWS.S3.Prefix
+	}
+
+	key := fmt.Sprintf("%s/%s/%s.json", prefix, t.Format("2006-01-02"), t.Format("2006-01-02"))
```
Comment (Issif) [inline outputs/aws.go:151]:
```suggestion
	key := fmt.Sprintf("%s/%s/%s.json", prefix, t.Format("2006-01-02"), t.Format(time.RFC3339Nano))
```

File: outputs/aws.go [hunk 9]
```diff
@@ -135,6 +138,38 @@ func (c *Client) SendMessage(falcopayload types.FalcoPayload) {
...
+	})
+	if err != nil {
+		go c.CountMetric("outputs", 1, []string{"output:awss3", "status:error"})
+		c.PromStats.Outputs.With(map[string]string{"destination": "awss3", "status": Error}).Inc()
+		log.Printf("[ERROR] : %v S3 - %v\n", c.OutputType, err.Error())
+		return
+	}
+
+	if c.Config.Debug == true {
+		log.Printf("[DEBUG] : %v S3 - SSECustomerKeyMD5 : %v\n", c.OutputType, *resp.SSECustomerKeyMD5)
```
Comment (Issif) [inline outputs/aws.go:165]: I'm not sure it's useful as we print the md5 in anycase (Line 168)

**Review by Issif — CHANGES_REQUESTED** [review]

**Review by Issif — COMMENTED** [review]
It seems OK for me, I'll make some tests before approving.

Can you update the `README` et `config_example.yaml`, please?

File: README.md [hunk 10]
```diff
@@ -223,6 +223,10 @@ aws:
     # loggroup : "" #  AWS CloudWatch Logs Group name, if not empty, CloudWatch Logs output is enabled
     # logstream : "" # AWS CloudWatch Logs Stream name, if empty, Falcosidekick will try to create a log stream
     # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+  s3:
+    # bucket: "falcosidekick" # AWS S3, bucket name
+    # prefix : "" # name of prefix
```
Comment (Issif) [inline README.md:228]:
```suggestion
    # prefix : "" # name of prefix, keys will have format: s3://<bucket>/<prefix>/YYYY-MM-DD/YYYY-MM-DDTHH:mm:ss.s+01:00.json
```

File: README.md [hunk 11]
```diff
@@ -459,6 +463,11 @@ care of lower/uppercases**) : `yaml: a.b --> envvar: A_B` :
 - **AWS_CLOUDWATCHLOGS_MINIMUMPRIORITY** : minimum priority of event for using
   this output, order is
   `emergency|alert|critical|error|warning|notice|informational|debug or "" (default)`
+- **AWS_S3_BUCKET** : AWS S3 Bucket, if not empty, AWS S3 output is
+    _enabled_
+- **AWS_S3_PREFIX** : Prefix name of the object
```
Comment (Issif) [inline README.md:468]:
```suggestion
- **AWS_S3_PREFIX** : Prefix name of the object, keys will have format: s3://<bucket>/<prefix>/YYYY-MM-DD/YYYY-MM-DDTHH:mm:ss.s+01:00.json
```

**Review by Issif — CHANGES_REQUESTED** [review]
I added some comments, and it remains `config_example.yaml` to update please

**Review by Issif — CHANGES_REQUESTED** [review]
Just add AWS S3 in the list of available outputs in README and we're good 👍

**Comment by evalsocket** [comment]
@Issif README changes are done

**Comment by cpanato** [comment]
/lgtm

thanks for this addition

**Comment by Issif** [comment]
> @Issif README changes are done

I can't see AWS S3 in List of available outputs 🤔

**Review by Issif — APPROVED** [review]

**Pull request merged**